terrain_simplification/src/terrain_simplification.cpp

```diff
@@ -0,0 +1,258 @@
+  mutex_state_.unlock();
+  map_sub_ = map_full_.getSubmap(robot_position, grid_map::Length(map_size_.x(), map_size_.y()), success);
+
+  // Create empty map
+  grid_map::GridMap map_simplified_wo_traversability({"simplified"});
+  map_simplified_wo_traversability.setFrameId("point_cloud_odom");
+  map_simplified_wo_traversability.setGeometry(map_sub_.getLength(), 0.02);
+  map_simplified_wo_traversability.setPosition(robot_position);
+
+  // Copy orignal layers
+  std::vector<std::string> layers = {"elevation"};
+  if (!map_simplified_wo_traversability.addDataFrom(map_sub_, true, true, true, layers)) {
+    ROS_ERROR("Could not add data from map_sub_ to map_filtered_.");
+    return;
+  }
+
+  // Simplify the terrain
+  convertGridMapToCvImage("elevation", map_sub_, img_raw_);             // takes 0.5 ms
+  mutex_.lock(); // to write and img_simplified_
+  filterCvImage(img_raw_, img_simplified_);                             // takes 1-10 ms
+  convertCvImageToGridMap("simplified", img_simplified_, map_simplified_wo_traversability); // takes 1-2 ms
+  mutex_.unlock();
+
+  // Apply the filter chain to create a separate map with a traversability layer
+  grid_map::GridMap map_simplified({"simplified"});
+  applyFilterChain(map_simplified_wo_traversability, map_simplified);                       // takes ~100-200 ms
+
+  mutex_.lock(); // to write map_filtered_ and map_simplified_
+  map_simplified_wo_traversability_ = map_simplified_wo_traversability;
+  map_simplified_ = map_simplified;
+  ready_ = true;
+  mutex_.unlock();
+}
+
+void
+TerrainSimplification::convertGridMapToCvImage (
+    const std::string& layer_name,
+    const grid_map::GridMap& map,
+    cv::Mat& image) {
+  grid_map::GridMapCvConverter::toImage<unsigned short, 1>(
+        map, layer_name, CV_16UC1, 0.0, 1.0, image);
+
+}
+
+void
+TerrainSimplification::convertCvImageToGridMap (
+    const std::string& layer_name,
+    const cv::Mat& image,
+    grid_map::GridMap& map) {
+  grid_map::GridMapCvConverter::addLayerFromImage<unsigned short, 1>(
+        image, layer_name, map);
+}
+
+void
+TerrainSimplification::filterCvImage (
+    const cv::Mat& image,
+    cv::Mat& image_filtered) {
+
+  cv::Mat image_median; cv::medianBlur(image, image_median, 5);
+
+  // not used
+  //  cv::Mat image_gaussian;
+  //  int size_x = abs(2*floor((sin(getRobotYaw())*51.)/2.)+1);
+  //  int size_y = abs(2*floor((cos(getRobotYaw())*51.)/2.)+1);
+  //  cv::GaussianBlur(image_median, image_gaussian, cv::Size(size_x, size_y), 0.0, 0.0);
+
+  cv::Mat image_directional;
+  cv::filter2D(image_median, image_directional, -1, getDirectionalBlurKernel());
+
+  image_filtered = image_directional;
+}
+
+cv::Mat
+TerrainSimplification::getDirectionalBlurKernel() {
+  cv::Mat k;
+  k = cv::Mat::zeros(k_size_, k_size_, CV_64F);
+  k.row((int)(k_size_-1)/2) = cv::getGaussianKernel(k_size_, -1, CV_64F).t();
+  double yaw = getRobotYaw();
+  cv::warpAffine(
```

Comment:
weird formatting

Reply:
I thought it is impossible to understand if it is in-line because of one of the inputs being `cv::getRotationMatrix2D()`. Should I:
- keep it, or
- have everything in a single line, or
- put cv::getRotationMatrix2D()on a separate line, or
- compute cv::getRotationMatrix2D() before `cv::warpAffine()`?

Reply:
you can keep it, or put the arguments in two or three lines